### PR TITLE
Add dilate option and MultiNodeBatchNormalization to Conv2DActiv and Conv2DBNActiv

### PR DESCRIPTION
**Pull request by mitmul**
#388 became a too large PR, then I split it into some small PRs.
First, `Conv2DActiv` and `Conv2DBNActive` should be able to take `dilate` option to use `DilatedConvolution2D` instead of `Convolution2D`.
Next, `Conv2DBNActive` should be able to take [`chainermn.links.MultiNodeBatchNormalization`](http://chainermn.readthedocs.io/en/stable/reference/index.html#chainermn.links.MultiNodeBatchNormalization) to use batch normalization layers correctly with multiple GPUs.
This PR introduces these two features.

### CODE REVIEW

File: chainercv/links/connection/conv_2d_bn_activ.py [hunk 1]
```diff
@@ -12,7 +19,10 @@ class Conv2DBNActiv(chainer.Chain):
 
     The arguments are the same as that of
     :class:`chainer.links.Convolution2D`
-    except for :obj:`activ` and :obj:`bn_kwargs`.
+    except for :obj:`activ`, :obj:`bn_kwargs`, and :obj:`comm`.
+    :obj:`comm` is a communicator of ChainerMN which is used for
+    :obj:`MultiNodeBatchNormalization`. If :obj:`None` is given to the argument
```
Comment (Hakuyume) [inline chainercv/links/connection/conv_2d_bn_activ.py]:
```
:class:`chainermn.links.MultiNodeBatchNormalization`
```

File: chainercv/links/connection/conv_2d_bn_activ.py [hunk 2]
```diff
@@ -57,22 +69,36 @@ class Conv2DBNActiv(chainer.Chain):
             :func:`chainer.functions.relu`.
         bn_kwargs (dict): Keyword arguments passed to initialize
             :class:`chainer.links.BatchNormalization`.
+        comm (:class:`~chainermn.communicators.CommunicatorBase):
+            If a ChainerMN communicator is given,
+            :obj:`~chainermn.links.MultiNodeBatchNormalization` will be used
+            for the batch normalization. If :obj:`None`,
+            :obj:`~chainer.links.BatchNormalization` will be used.
```
Comment (Hakuyume) [inline chainercv/links/connection/conv_2d_bn_activ.py]: I prefer to take  `comm` as an element of `bn_kwargs` because `comm` is used only for batchnorm.
If `'comm' in bn_kwargs`, it use `MultiNodeBatchNormalization`.

File: chainercv/links/connection/conv_2d_bn_activ.py (hunk 1 above)
Comment (mitmul) [reply chainercv/links/connection/conv_2d_bn_activ.py]: Fixed

File: chainercv/links/connection/conv_2d_bn_activ.py (hunk 2 above)
Comment (mitmul) [reply chainercv/links/connection/conv_2d_bn_activ.py]: OK, I changed it

File: chainercv/links/connection/conv_2d_bn_activ.py [hunk 3]
```diff
@@ -21,8 +21,8 @@ class Conv2DBNActiv(chainer.Chain):
     :class:`chainer.links.Convolution2D`
     except for :obj:`activ`, :obj:`bn_kwargs`, and :obj:`comm`.
     :obj:`comm` is a communicator of ChainerMN which is used for
-    :obj:`MultiNodeBatchNormalization`. If :obj:`None` is given to the argument
-    :obj:`comm`, :obj:`BatchNormalization` link from Chainer is used.
+    :class:`chainermn.links.MultiNodeBatchNormalization`. If
+    :obj:`None` is given to the argument :obj:`comm`, :obj:`BatchNormalization` link from Chainer is used.
```
Comment (Hakuyume) [inline chainercv/links/connection/conv_2d_bn_activ.py]:
```
:class:`chainer.links.BatchNormalization`
```

File: chainercv/links/connection/conv_2d_bn_activ.py [hunk 4]
```diff
@@ -94,9 +94,9 @@ def __init__(self, in_channels, out_channels, ksize=None,
                 self.conv = Convolution2D(
                     in_channels, out_channels, ksize, stride, pad,
                     nobias, initialW, initial_bias)
-            if comm is not None and _chainermn_available:
+            if 'comm' in bn_kwargs and _chainermn_available:
                 self.bn = MultiNodeBatchNormalization(
-                    out_channels, comm, **bn_kwargs)
+                    out_channels, [**bn_kwargs)
```
Comment (Hakuyume) [inline chainercv/links/connection/conv_2d_bn_activ.py]: `[` ?

File: chainercv/links/connection/conv_2d_bn_activ.py [hunk 5]
```diff
@@ -68,11 +68,11 @@ class Conv2DBNActiv(chainer.Chain):
         activ (callable): An activation function. The default value is
             :func:`chainer.functions.relu`.
         bn_kwargs (dict): Keyword arguments passed to initialize
-            :class:`chainer.links.BatchNormalization`.
-        comm (:class:`~chainermn.communicators.CommunicatorBase):
-            If a ChainerMN communicator is given,
+            :class:`chainer.links.BatchNormalization`. If a ChainerMN
+            communicator (:class:`~chainermn.communicators.CommunicatorBase)
+            is given with the key :obj:`comm`,
             :obj:`~chainermn.links.MultiNodeBatchNormalization` will be used
```
Comment (Hakuyume) [inline chainercv/links/connection/conv_2d_bn_activ.py]:
```
:class:`chainermn.links.MultiNodeBatchNormalization`
```

File: chainercv/links/connection/conv_2d_bn_activ.py [hunk 6]
```diff
@@ -2,6 +2,13 @@
 from chainer.functions import relu
 from chainer.links import BatchNormalization
 from chainer.links import Convolution2D
+from chainer.links import DilatedConvolution2D
+
+try:
+    from chainermn.links import MultiNodeBatchNormalization
+    _chainermn_available = True
+except (ImportError, TypeError):
```
Comment (yuyu2172) [inline chainercv/links/connection/conv_2d_bn_activ.py]: Does TypeError occur?
If not, we can remove it.

File: chainercv/links/connection/conv_2d_bn_activ.py [hunk 7]
```diff
@@ -56,23 +70,37 @@ class Conv2DBNActiv(chainer.Chain):
...
+            for the batch normalization. Otherwise,
+            :obj:`~chainer.links.BatchNormalization` will be used.
 
     """
 
     def __init__(self, in_channels, out_channels, ksize=None,
-                 stride=1, pad=0, nobias=True, initialW=None,
-                 initial_bias=None, activ=relu, bn_kwargs=dict()):
+                 stride=1, pad=0, dilate=1, nobias=True, initialW=None,
+                 initial_bias=None, activ=relu, bn_kwargs=dict(), comm=None):
```
Comment (yuyu2172) [inline chainercv/links/connection/conv_2d_bn_activ.py]: Can you remove `comm` option?

File: tests/links_tests/connection_tests/test_conv_2d_bn_activ.py [hunk 8]
```diff
@@ -104,4 +113,72 @@ def test_backward_gpu(self):
...
+                            dtype=np.float32).reshape((1, 1, 3, 3))
+        bn_kwargs = {'decay': 0.8}
+        initial_bias = 0
+        comm = create_communicator('naive')
+        activ = relu
+        self.l = Conv2DBNActiv(
+            self.in_channels, self.out_channels, self.ksize, self.stride,
+            self.pad, self.dilate, initialW=initialW,
+            initial_bias=initial_bias, activ=activ, bn_kwargs=bn_kwargs,
+            comm=comm)
```
Comment (yuyu2172) [inline tests/links_tests/connection_tests/test_conv_2d_bn_activ.py]: Include `comm` in `bn_kwards`.

**Comment by yuyu2172** [comment]
I am very sorry for late review.
Could you please resolve the conflict with master branch?

File: .travis.yml [hunk 9]
```diff
@@ -28,8 +28,15 @@ install:
   - conda info -a
 
   - if [[ "$OPTIONAL_MODULES" == "1" ]]; then
+      export LIBRARY_PATH="$HOME/miniconda/lib:$LIBRARY_PATH";
       conda env create -f environment.yml;
       source activate chainercv;
+      cd $HOME;
+      wget https://github.com/chainer/chainermn/archive/v1.0.0.tar.gz -O chainermn.tar.gz;
```
Comment (yuyu2172) [inline .travis.yml]: Please update to the latest release.

File: tests/links_tests/connection_tests/test_conv_2d_bn_activ.py [hunk 10]
```diff
@@ -15,7 +15,15 @@ def _add_one(x):
     return x + 1
 
 
+try:
+    from chainermn import create_communicator
+    _chainermn_available = True
+except (ImportError, TypeError):
```
Comment (yuyu2172) [inline tests/links_tests/connection_tests/test_conv_2d_bn_activ.py]: Remove TypeError if not necessary.

**Comment by mitmul** [comment]
@yuyu2172 Thanks for the review. I fixed all the points commented in review.

**Comment by yuyu2172** [comment]
Thank you. Please fix tests.

File: tests/links_tests/connection_tests/test_conv_2d_activ.py [hunk 11]
```diff
@@ -93,7 +94,7 @@ def test_forward_gpu(self):
     def check_backward(self, x_data, y_grad):
         x = chainer.Variable(x_data)
         y = self.l(x)
-        y.grad = y_grad
+        y.grad = y_grad if self.dilate == 1 else y_grad[:, :, 1:-1, 1:-1]
```
Comment (yuyu2172) [inline tests/links_tests/connection_tests/test_conv_2d_activ.py]: How about

```
if self.dilate == 1:
     y.grad = y_grad
elif self.dilate == 2:
     y.grad = y_grad[:, :, 1:-1, 1:-1]
```

This is better because ` y_grad[:, :, 1:-1, 1:-1]` would not work when `dilate > 2`.

File: tests/links_tests/connection_tests/test_conv_2d_bn_activ.py [hunk 12]
```diff
@@ -72,14 +80,15 @@ def check_forward(self, x_data):
         self.assertIsInstance(y, chainer.Variable)
         self.assertIsInstance(y.array, self.l.xp.ndarray)
 
+        _x_data = x_data if self.dilate == 1 else x_data[:, :, 1:-1, 1:-1]
```
Comment (yuyu2172) [inline tests/links_tests/connection_tests/test_conv_2d_bn_activ.py]: ditto

File: tests/links_tests/connection_tests/test_conv_2d_bn_activ.py [hunk 13]
```diff
@@ -99,7 +108,7 @@ def test_forward_gpu(self):
     def check_backward(self, x_data, y_grad):
         x = chainer.Variable(x_data)
         y = self.l(x)
-        y.grad = y_grad
+        y.grad = y_grad if self.dilate == 1 else y_grad[:, :, 1:-1, 1:-1]
```
Comment (yuyu2172) [inline tests/links_tests/connection_tests/test_conv_2d_bn_activ.py]: ditto

**Comment by yuyu2172** [comment]
ChainerMN seems to be not installed correctly.

It is raising an error `error: option --no-nccl not recognized`.

File: tests/links_tests/connection_tests/test_conv_2d_activ.py (hunk 11 above)
Comment (mitmul) [reply tests/links_tests/connection_tests/test_conv_2d_activ.py]: OK, I fixed it, but there's no test case here for dilate > 2.

**Comment by Hakuyume** [comment]
Why do you install ChainerMN from source? Can't we install it by modifying `chainercv/environment.yml`?

**Comment by yuyu2172** [comment]
Travis is failing because environment file is incorrect.
Perhaps, pip should be used inside the env file to install mn.

**Comment by yuyu2172** [comment]
@mitmul 

You need `-` before `pip`.

The env file below worked.

```
name: chainercv
channels:
- !!python/unicode
  'menpo'
- !!python/unicode
  'mpi4py'
- !!python/unicode
  'defaults'
dependencies:
  - Cython
  - opencv3=3.2.0
  - matplotlib
  - numpy
  - Pillow
  - openmpi
  - pip:
    - chainermn==1.2.0
```

**Comment by mitmul** [comment]
Thanks. I updated it.

**Comment by yuyu2172** [comment]
Tests are failing when `dilate == 2`.
I think you forgot to modify `test_conv_2d_activ.py`. It is fine for `test_conv_2d_bn_activ.py`.

**Comment by mitmul** [comment]
@yuyu2172 Sorry, I updated the test too.

**Comment by mitmul** [comment]
All tests passed.

**Comment by yuyu2172** [comment]
I checked the test log and it seems that ChainerMN is not installed properly.

The log says that `mpi4py` fails to get installed.
Could you fix this?

```
  _configtest.c:2:17: fatal error: mpi.h: No such file or directory
   #include <mpi.h>
                   ^
  compilation terminated.
  failure.
  removing: _configtest.c _configtest.o
  error: Cannot compile MPI programs. Check your configuration!!!
  
  ----------------------------------------
[31m  Failed building wheel for mpi4py[0m
[?25h  Running setup.py clean for mpi4py
  Running setup.py bdist_wheel for pycparser ... [?25l- \ done
[?25h  Stored in directory: /home/travis/.cache/pip/wheels/95/14/9a/5e7b9024459d2a6600aaa64e0ba485325aff7a9ac7489db1b6
  Running setup.py bdist_wheel for filelock ... [?25l- done
[?25h  Stored in directory: /home/travis/.cache/pip/wheels/5f/5e/8a/9f1eb481ffbfff95d5f550570c1dbeff3c1785c8383c12c62b
Successfully built chainermn chainer pycparser filelock
Failed to build mpi4py
```

As a consequence, the tests related to ChainerMN are skipped.
https://travis-ci.org/chainer/chainercv/jobs/354200398#L1077

**Comment by mitmul** [comment]
I see. Thanks for catching it. I'll fix that.

**Comment by mitmul** [comment]
@yuyu2172 Fixed the `environment.yml` to install `mpi4py` from conda. I think it fixed the installation of mpi4py.
